tuna/miopen/celery_tuning/tuning.py

```diff
@@ -0,0 +1,395 @@
+        set_job_state(session,
+                      job,
+                      dbt,
+                      orig_state,
+                      increment_retries=True,
+                      result=result_str)
+    else:
+      LOGGER.info("\n\n Setting job state to evaluated")
+      set_job_state(session, job, dbt, 'evaluated', result=result_str)
+      clean_cache_table(dbt, job)
+  except (OperationalError, IntegrityError) as err:
+    LOGGER.warning('FinBuild: Unable to update Database %s', err)
+    session.rollback()
+    failed_job = True
+
+  return True
+
+
+def set_job_state(session, job, dbt, state, increment_retries=False, result=""):
+  """Interface function to update job state for builder/evaluator
+  job_set_attr: List[str]"""
+
+  LOGGER.info('Setting job id %s state to %s', job.id, state)
+  job_set_attr = ['state', 'gpu_id']
+  job.state = state
+  if result:
+    job_set_attr.append('result')
+    job.result = result
+  if increment_retries:
+    job_set_attr.append('retries')
+    job.retries += 1
+
+  #pylint: disable=duplicate-code
+  if '_start' in state:
+    job_set_attr.append('cache_loc')
+    cache: str = '~/.cache/miopen_'
+    blurr: str = ''.join(
+        random.choice(string.ascii_lowercase) for i in range(10))
+    cache_loc: str = cache + blurr
+    job.cache_loc = cache_loc
+  #pylint: enable=duplicate-code
+
+  query: str = gen_update_query(job, job_set_attr, dbt.job_table.__tablename__)
+
+  def callback() -> bool:
+    session.execute(query)
+    session.commit()
+    return True
+
+  assert session_retry(session, callback, lambda x: x(), LOGGER)
+  return True
+
+
+def get_worker_granularity(library):
+  """Check how many celery workers we need"""
+  worker_granularity = None
+  worker_granularity = get_worker_type(library.args)
+  if 'miopen_find_compile' in library.args.fin_steps \
+  or 'miopen_perf_compile' in library.args.fin_steps:
+    worker_granularity = 'worker_per_node'
+  elif 'miopen_find_eval' in library.args.fin_steps or 'miopen_perf_eval' in library.args.fin_steps:
+    worker_granularity = 'worker_per_gpu'
+
+  return worker_granularity
+
+
+def prep_tuning(library):
+  """Prep env for tuning start"""
+  worker_type = get_worker_type(library.args)
+  machines = load_machines(library.args)
+  q_name = None
+  if worker_type == 'fin_build_worker':
+    q_name = f"compile_q_session_{library.dbt.session_id}"
+  else:
+    q_name = f"eval_q_session_{library.dbt.session_id}"
+
+  if not library.args.enqueue_only:
+    try:
+      #stop_active_workers()
+      if not launch_celery_worker(library, q_name, machines,
+                                  get_worker_granularity(library)):
+        raise ValueError('Could not launch celery worker')
+    except kombu.exceptions.OperationalError as k_err:
+      LOGGER.error('Redis error ocurred: %s', k_err)
+      return False
+
+  global DBT  #pylint: disable=global-variable-undefined
+  DBT = MIOpenDBTables(session_id=library.args.session_id,
```

Comment:
i tried, i did not find a way to get rid of it. Information from here is needed before and after celery tasks have run and since this structure is not serializable, we cannot pass it to celery.

Reply:
I see.